homeassistant/components/dreo/fan.py

```diff
@@ -0,0 +1,155 @@
+        return None
+
+    @cached_property
+    def speed_count(self) -> int:
+        """Return the number of speeds the fan supports."""
+        if hasattr(self, "_attr_speed_count"):
+            return self._attr_speed_count
+        return 6
+
+    @cached_property
+    def percentage(self) -> int | None:
+        """Return the current speed as a percentage."""
+        if hasattr(self, "_attr_percentage"):
+            return ranged_value_to_percentage(
+                FAN_DEVICE.get("config").get(self._model).get("speed_range"),
+                self._attr_percentage
+            )
+        return 0
+
+    @property
+    def available(self) -> bool:
+        """Return True if device is available."""
+        return self._attr_available
+
+    def turn_on(self, **kwargs: Any) -> None:
+        """Turn the device on."""
+        self._try_command(
+            "Turn the device on failed.", power_switch=True
+        )
+        self._attr_state = True
+
+    def turn_off(self, **kwargs: Any) -> None:
+        """Turn the device off."""
+        self._try_command(
+            "Turn the device off failed.", power_switch=False
+        )
+        self._attr_state = False
+
+    def set_preset_mode(self, preset_mode: str) -> None:
+        """Set the preset mode of fan."""
+        self._try_command(
+            "Set the preset mode failed.", mode=preset_mode
+        )
+        self._attr_preset_mode = preset_mode
+
+    def set_percentage(self, percentage: int) -> None:
+        """Set the speed of fan."""
+        speed = math.ceil(
+            percentage_to_ranged_value(
+                self._attr_low_high_range, percentage
+            )
+        )
+
+        self._try_command(
+            "Set the speed failed.", speed=speed
+        )
+        self._attr_percentage = speed
+
+    def oscillate(self, oscillating: bool) -> None:
+        """Set the Oscillate of fan."""
+        self._try_command(
+            "Set the Oscillate failed.", oscillate=oscillating
+        )
+        self._attr_oscillating = oscillating
+
+    def update(self) -> None:
+        """Update Dreo fan."""
+        try:
+            status = self._config_entry.runtime_data.client.get_status(self._device_id)
+
+        except HsCloudException as ex:
+            _LOGGER.exception("unable to connect")
+            raise ConfigEntryNotReady(f"unable to connect") from ex
+
+        except HsCloudBusinessException as ex:
+            _LOGGER.exception("invalid username or password")
+            raise ConfigEntryNotReady(f"invalid username or password") from ex
+
+        except Exception as ex:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            raise ConfigEntryNotReady(f"Unexpected exception") from ex
+
+        if status is not None:
+            self._attr_state = status.get('power_switch')
+            self._attr_preset_mode = status.get('mode')
+            self._attr_percentage = status.get('speed')
```

Comment:
If I get it correct `status.get("speed")` gives a range value. I think you should convert it to a percentage here.

```suggestion
            self._attr_percentage = ranged_value_to_percentage(
                self._attr_low_high_range,
                status.get('speed'),
            )
```